### PR TITLE
Improvements to `GaussianConditional` and `GaussianBayesNet`

### PR DESCRIPTION
**Pull request by varunagrawal**
1. `GaussianBayesNet::logNormalizationConstant` method.
2. Equality operators for both `GaussianBayesNet` and `GaussianConditional`.
3. Print the `logNormalizationConstant` when printing the `GaussianConditional`.
4. Wrap more `GaussianConditional` constructors.

### CODE REVIEW

File: gtsam/linear/GaussianBayesNet.cpp
Comment (dellaert) [inline gtsam/linear/GaussianBayesNet.cpp:247]: Is this copy/paste from gc ?

File: gtsam/linear/GaussianBayesNet.h
Comment (dellaert) [inline gtsam/linear/GaussianBayesNet.h:88]: But this is not exact equality

File: gtsam/linear/VectorValues.h
Comment (dellaert) [inline gtsam/linear/VectorValues.h:268]: But this is not exact….

File: gtsam/linear/GaussianBayesNet.h
Comment (dellaert) [inline gtsam/linear/GaussianBayesNet.h:243]: No unit test

**Review by dellaert — COMMENTED** [review]
Several comments. Also puzzled why not targeting develop. I want to limit the amount of change in the other branch to ease review, and this seems completely orthogonal.

File: gtsam/linear/GaussianBayesNet.cpp
Comment (varunagrawal) [reply gtsam/linear/GaussianBayesNet.cpp:247]: Not quite. It collects all the `logNormalizationConstants` from the GCs.

File: gtsam/linear/GaussianBayesNet.h
Comment (varunagrawal) [reply gtsam/linear/GaussianBayesNet.h:88]: The docstring for the `isEqual` method in `FactorGraph.h` says
```
/// Check exact equality of the factor pointers. Useful for derived ==.
```
You wrote that docstring 3 years ago.

Comment (dellaert) [reply gtsam/linear/GaussianBayesNet.h:88]: Ah, isEqual might be. Please check it. `equals`is not.

**Comment by varunagrawal** [comment]
> Also puzzled why not targeting develop. I want to limit the amount of change in the other branch to ease review, and this seems completely orthogonal.

This was all part of 1 PR initially and I broke it down. Chaining the PRs this way allows me to create a PR on top of both PRs rather than just using the changes in either one.

I don't not target to `develop` on purpose.

File: gtsam/linear/GaussianBayesNet.h
Comment (varunagrawal) [reply gtsam/linear/GaussianBayesNet.h:243]: Good catch. I'm updating a unit test in `testGaussianBayesNet.cpp` that will test this.

File: gtsam/linear/GaussianBayesNet.h
Comment (varunagrawal) [reply gtsam/linear/GaussianBayesNet.h:88]: Yup, I edited the other one which uses `equals`. 🙂

**Comment by dellaert** [comment]
> > Also puzzled why not targeting develop. I want to limit the amount of change in the other branch to ease review, and this seems completely orthogonal.
> 
> This was all part of 1 PR initially and I broke it down. Chaining the PRs this way allows me to create a PR on top of both PRs rather than just using the changes in either one.
> 
> I don't not target to `develop` on purpose.

Ok, so the intent is that the base will change to develop?

**Comment by varunagrawal** [comment]
Exactly. Once I merge the ADT PR and delete that branch, GitHub will rebase this PR to develop automatically. :)

File: gtsam/linear/VectorValues.h
Comment (dellaert) [reply gtsam/linear/VectorValues.h:268]: Still uses equals, no?

**Review by dellaert — APPROVED** [review]
Great. Still one issue - equals instead of isEqual in ==?

File: gtsam/linear/VectorValues.h
Comment (varunagrawal) [reply gtsam/linear/VectorValues.h:268]: I edited the docstring and it says `/// Check equality.` Is that not showing for you?

Comment (dellaert) [reply gtsam/linear/VectorValues.h:268]: operator== implies strict equality. If you want equals, then please rename to `equals` and provide a tolerance parameter.

**Comment by varunagrawal** [comment]
> Great. Still one issue - equals instead of isEqual in ==?

I updated the docstring to say `Check equality` instead of `Check exact equality` immediately after your review comment. I am going to merge this, and if you have another comment, I can add it to the next PR in this series.

**Comment by dellaert** [comment]
Yes, I think you missed my comment:
“operator== implies strict equality. If you want equals, then please rename to equals and provide a tolerance parameter.”

I feel strongly about being consistent across the code base here. If we find other instances of this we should also correct those.

**Pull request merged**
